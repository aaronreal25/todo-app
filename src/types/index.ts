export type TaskStatus = 'PENDING' | 'IN_PROGRESS' | 'COMPLETED';

export interface Task {
  id: string;
  title: string;
  description: string;
  status: TaskStatus;
  dueDate: string;
  createdAt: string;
}

export interface TaskFilter {
  status: TaskStatus | 'ALL';
  searchQuery: string;
}

export interface TaskState {
  tasks: Task[];
  filter: TaskFilter;
  isLoading: boolean;
  error: string | null;
}

export type TaskAction =
  | { type: 'SET_TASKS'; payload: Task[] }
  | { type: 'ADD_TASK'; payload: Task }
  | { type: 'UPDATE_TASK'; payload: Task }
  | { type: 'DELETE_TASK'; payload: string }
  | { type: 'SET_FILTER'; payload: Partial<TaskFilter> }
  | { type: 'SET_LOADING'; payload: boolean }
  | { type: 'SET_ERROR'; payload: string | null };