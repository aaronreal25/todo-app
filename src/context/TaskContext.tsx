'use client';

import React, { createContext, useContext, useReducer, useCallback } from 'react';
import { Task, TaskStatus } from '@/types';

interface TaskState {
  tasks: Task[];
  isLoading: boolean;
  error: string | null;
}

type TaskAction =
  | { type: 'ADD_TASK'; payload: Task }
  | { type: 'UPDATE_TASK'; payload: Task }
  | { type: 'DELETE_TASK'; payload: string }
  | { type: 'SET_LOADING'; payload: boolean }
  | { type: 'SET_ERROR'; payload: string | null };

const initialState: TaskState = {
  tasks: [],
  isLoading: false,
  error: null,
};

function taskReducer(state: TaskState, action: TaskAction): TaskState {
  switch (action.type) {
    case 'ADD_TASK':
      return {
        ...state,
        tasks: [...state.tasks, action.payload],
        isLoading: false,
        error: null,
      };
    
    case 'UPDATE_TASK':
      return {
        ...state,
        tasks: state.tasks.map((task) =>
          task.id === action.payload.id ? action.payload : task
        ),
      };
    
    case 'DELETE_TASK':
      return {
        ...state,
        tasks: state.tasks.filter((task) => task.id !== action.payload),
      };
    
    case 'SET_LOADING':
      return {
        ...state,
        isLoading: action.payload,
      };
    
    case 'SET_ERROR':
      return {
        ...state,
        error: action.payload,
      };
    
    default:
      return state;
  }
}

// Export the context
export const TaskContext = createContext<{
  state: TaskState;
  addTask: (task: Omit<Task, 'id' | 'createdAt'>) => void;
  updateTask: (task: Task) => void;
  deleteTask: (id: string) => void;
} | null>(null);

export const TaskProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [state, dispatch] = useReducer(taskReducer, initialState);

  const addTask = useCallback((taskData: Omit<Task, 'id' | 'createdAt'>) => {
    const newTask: Task = {
      ...taskData,
      id: crypto.randomUUID(),
      createdAt: new Date().toISOString(),
    };
    dispatch({ type: 'ADD_TASK', payload: newTask });
  }, []);

  const updateTask = useCallback((task: Task) => {
    dispatch({ type: 'UPDATE_TASK', payload: task });
  }, []);

  const deleteTask = useCallback((id: string) => {
    dispatch({ type: 'DELETE_TASK', payload: id });
  }, []);

  return (
    <TaskContext.Provider 
      value={{ 
        state,
        addTask,
        updateTask,
        deleteTask,
      }}
    >
      {children}
    </TaskContext.Provider>
  );
};

export const useTaskContext = () => {
  const context = useContext(TaskContext);
  if (!context) {
    throw new Error('useTaskContext must be used within a TaskProvider');
  }
  return context;
};
