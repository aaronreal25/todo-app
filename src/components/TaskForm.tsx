'use client';

import React, { useState } from 'react';
import { Task, TaskStatus } from '@/types';
import { useTaskContext } from '@/context/TaskContext';

interface TaskFormProps {
  task?: Task;
  onClose: () => void;
}

const TaskForm: React.FC<TaskFormProps> = ({ task, onClose }) => {
  const { addTask, updateTask } = useTaskContext();
  const [formData, setFormData] = useState({
    title: task?.title || '',
    description: task?.description || '',
    status: task?.status || 'PENDING' as TaskStatus,
    dueDate: task?.dueDate || new Date().toISOString().split('T')[0],
  });

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    
    if (task) {
      updateTask({
        ...task,
        ...formData,
      });
    } else {
      addTask(formData);
    }
    
    onClose();
  };

  return (
    <form onSubmit={handleSubmit} className="task-form space-y-6">
      <div className="task-form-field">
        <label htmlFor="task-title" className="task-form-label block text-sm font-medium text-gray-200 mb-2">Title</label>
        <input
          id="task-title"
          type="text"
          value={formData.title}
          onChange={(e) => setFormData({ ...formData, title: e.target.value })}
          className="task-form-input-title mt-1 block w-full rounded-md bg-gray-700 border-gray-600 text-white focus:border-blue-500 focus:ring-blue-500 px-2 py-2"
          required
        />
      </div>

      <div className="task-form-field">
        <label htmlFor="task-description" className="task-form-label block text-sm font-medium text-gray-200 mb-2">Description</label>
        <textarea
          id="task-description"
          value={formData.description}
          onChange={(e) => setFormData({ ...formData, description: e.target.value })}
          className="task-form-input-description mt-1 block w-full rounded-md bg-gray-700 border-gray-600 text-white focus:border-blue-500 focus:ring-blue-500 px-2 py-2"
          rows={4}
          required
        />
      </div>

      <div className="task-form-field">
        <label htmlFor="task-status" className="task-form-label block text-sm font-medium text-gray-200 mb-2">Status</label>
        <select
        id="task-status"
          value={formData.status}
          onChange={(e) => setFormData({ ...formData, status: e.target.value as TaskStatus })}
          className="task-form-select-status mt-1 block w-full rounded-md bg-gray-700 border-gray-600 text-white focus:border-blue-500 focus:ring-blue-500 px-2 py-2"
        >
          <option value="PENDING">Pending</option>
          <option value="IN_PROGRESS">In Progress</option>
          <option value="COMPLETED">Completed</option>
        </select>
      </div>

      <div className="task-form-field">
        <label htmlFor="task-due-date" className="task-form-label block text-sm font-medium text-gray-200 mb-2">Due Date</label>
        <input
          id="task-due-date"
          type="date"
          value={formData.dueDate}
          onChange={(e) => setFormData({ ...formData, dueDate: e.target.value })}
          className="task-form-input-date mt-1 block w-full rounded-md bg-gray-700 border-gray-600 text-white focus:border-blue-500 focus:ring-blue-500 px-2 py-2"
          required
        />
      </div>

      <div className="task-form-actions flex justify-end space-x-3 mt-8">
        <button
          type="button"
          onClick={onClose}
          className="task-form-button-cancel px-2 py-2 text-sm font-medium text-gray-200 bg-gray-700 border border-gray-600 rounded-md hover:bg-gray-600"
        >
          Cancel
        </button>
        <button
          type="submit"
          className="task-form-button-submit px-2 py-2 text-sm font-medium text-white bg-blue-600 border border-transparent rounded-md hover:bg-blue-700"
        >
          {task ? 'Update Task' : 'Add Task'}
        </button>
      </div>
    </form>
  );
};

export default TaskForm;
