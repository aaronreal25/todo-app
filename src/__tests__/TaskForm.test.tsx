import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import TaskForm from '@/components/TaskForm';
import { TaskProvider, TaskContext } from '@/context/TaskContext';
import { Task, TaskStatus } from '@/types';

describe('TaskForm', () => {
  const mockOnClose = jest.fn();
  const mockAddTask = jest.fn();
  const mockUpdateTask = jest.fn();

  const mockContextValue = {
    state: { tasks: [], isLoading: false, error: null },
    addTask: mockAddTask,
    updateTask: mockUpdateTask,
    deleteTask: jest.fn()
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  // Render test
  it('renders empty form correctly', () => {
    render(
      <TaskContext.Provider value={mockContextValue}>
        <TaskForm onClose={mockOnClose} />
      </TaskContext.Provider>
    );

    expect(screen.getByLabelText(/title/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/description/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/status/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/due date/i)).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /add task/i })).toBeInTheDocument();
  });

  it('submits new task correctly', () => {
    render(
      <TaskContext.Provider value={mockContextValue}>
        <TaskForm onClose={mockOnClose} />
      </TaskContext.Provider>
    );

    // Fill out the form
    fireEvent.change(screen.getByLabelText(/title/i), { 
      target: { value: 'New Task' } 
    });
    fireEvent.change(screen.getByLabelText(/description/i), { 
      target: { value: 'New Description' } 
    });
    fireEvent.change(screen.getByLabelText(/status/i), { 
      target: { value: 'IN_PROGRESS' } 
    });
    fireEvent.change(screen.getByLabelText(/due date/i), { 
      target: { value: '2024-12-31' } 
    });

    // Submit the form
    fireEvent.click(screen.getByRole('button', { name: /add task/i }));

    // Check if addTask was called with correct data
    expect(mockAddTask).toHaveBeenCalledWith({
      title: 'New Task',
      description: 'New Description',
      status: 'IN_PROGRESS',
      dueDate: '2024-12-31'
    });
    expect(mockOnClose).toHaveBeenCalled();
  });

  it('updates existing task correctly', () => {
    const existingTask: Task = {
      id: '1',
      title: 'Existing Task',
      description: 'Existing Description',
      status: 'PENDING' as TaskStatus,
      dueDate: '2024-12-31',
      createdAt: '2024-01-01'
    };

    render(
      <TaskContext.Provider value={mockContextValue}>
        <TaskForm task={existingTask} onClose={mockOnClose} />
      </TaskContext.Provider>
    );

    // Verify form is pre-filled
    expect(screen.getByLabelText(/title/i)).toHaveValue('Existing Task');
    expect(screen.getByLabelText(/description/i)).toHaveValue('Existing Description');
    expect(screen.getByLabelText(/status/i)).toHaveValue('PENDING');
    expect(screen.getByLabelText(/due date/i)).toHaveValue('2024-12-31');

    // Update the task
    fireEvent.change(screen.getByLabelText(/title/i), { 
      target: { value: 'Updated Task' } 
    });
    fireEvent.click(screen.getByRole('button', { name: /update task/i }));

    // Check if updateTask was called with correct data
    expect(mockUpdateTask).toHaveBeenCalledWith({
      ...existingTask,
      title: 'Updated Task'
    });
    expect(mockOnClose).toHaveBeenCalled();
  });

  // Snapshot tests
  describe('Snapshots', () => {
    it('matches snapshot for empty form', () => {
      const { container } = render(
        <TaskContext.Provider value={mockContextValue}>
          <TaskForm onClose={mockOnClose} />
        </TaskContext.Provider>
      );
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot for edit form', () => {
      const existingTask: Task = {
        id: '1',
        title: 'Existing Task',
        description: 'Existing Description',
        status: 'PENDING' as TaskStatus,
        dueDate: '2024-12-31',
        createdAt: '2024-01-01'
      };

      const { container } = render(
        <TaskContext.Provider value={mockContextValue}>
          <TaskForm task={existingTask} onClose={mockOnClose} />
        </TaskContext.Provider>
      );
      expect(container).toMatchSnapshot();
    });
  });
});