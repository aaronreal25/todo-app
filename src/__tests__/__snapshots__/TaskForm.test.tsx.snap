// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskForm Snapshots matches snapshot for edit form 1`] = `
<div>
  <form
    class="task-form space-y-6"
  >
    <div
      class="task-form-field"
    >
      <label
        class="task-form-label block text-sm font-medium text-gray-200 mb-2"
        for="task-title"
      >
        Title
      </label>
      <input
        class="task-form-input-title mt-1 block w-full rounded-md bg-gray-700 border-gray-600 text-white focus:border-blue-500 focus:ring-blue-500 px-2 py-2"
        id="task-title"
        required=""
        type="text"
        value="Existing Task"
      />
    </div>
    <div
      class="task-form-field"
    >
      <label
        class="task-form-label block text-sm font-medium text-gray-200 mb-2"
        for="task-description"
      >
        Description
      </label>
      <textarea
        class="task-form-input-description mt-1 block w-full rounded-md bg-gray-700 border-gray-600 text-white focus:border-blue-500 focus:ring-blue-500 px-2 py-2"
        id="task-description"
        required=""
        rows="4"
      >
        Existing Description
      </textarea>
    </div>
    <div
      class="task-form-field"
    >
      <label
        class="task-form-label block text-sm font-medium text-gray-200 mb-2"
        for="task-status"
      >
        Status
      </label>
      <select
        class="task-form-select-status mt-1 block w-full rounded-md bg-gray-700 border-gray-600 text-white focus:border-blue-500 focus:ring-blue-500 px-2 py-2"
        id="task-status"
      >
        <option
          value="PENDING"
        >
          Pending
        </option>
        <option
          value="IN_PROGRESS"
        >
          In Progress
        </option>
        <option
          value="COMPLETED"
        >
          Completed
        </option>
      </select>
    </div>
    <div
      class="task-form-field"
    >
      <label
        class="task-form-label block text-sm font-medium text-gray-200 mb-2"
        for="task-due-date"
      >
        Due Date
      </label>
      <input
        class="task-form-input-date mt-1 block w-full rounded-md bg-gray-700 border-gray-600 text-white focus:border-blue-500 focus:ring-blue-500 px-2 py-2"
        id="task-due-date"
        required=""
        type="date"
        value="2024-12-31"
      />
    </div>
    <div
      class="task-form-actions flex justify-end space-x-3 mt-8"
    >
      <button
        class="task-form-button-cancel px-2 py-2 text-sm font-medium text-gray-200 bg-gray-700 border border-gray-600 rounded-md hover:bg-gray-600"
        type="button"
      >
        Cancel
      </button>
      <button
        class="task-form-button-submit px-2 py-2 text-sm font-medium text-white bg-blue-600 border border-transparent rounded-md hover:bg-blue-700"
        type="submit"
      >
        Update Task
      </button>
    </div>
  </form>
</div>
`;

exports[`TaskForm Snapshots matches snapshot for empty form 1`] = `
<div>
  <form
    class="task-form space-y-6"
  >
    <div
      class="task-form-field"
    >
      <label
        class="task-form-label block text-sm font-medium text-gray-200 mb-2"
        for="task-title"
      >
        Title
      </label>
      <input
        class="task-form-input-title mt-1 block w-full rounded-md bg-gray-700 border-gray-600 text-white focus:border-blue-500 focus:ring-blue-500 px-2 py-2"
        id="task-title"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="task-form-field"
    >
      <label
        class="task-form-label block text-sm font-medium text-gray-200 mb-2"
        for="task-description"
      >
        Description
      </label>
      <textarea
        class="task-form-input-description mt-1 block w-full rounded-md bg-gray-700 border-gray-600 text-white focus:border-blue-500 focus:ring-blue-500 px-2 py-2"
        id="task-description"
        required=""
        rows="4"
      />
    </div>
    <div
      class="task-form-field"
    >
      <label
        class="task-form-label block text-sm font-medium text-gray-200 mb-2"
        for="task-status"
      >
        Status
      </label>
      <select
        class="task-form-select-status mt-1 block w-full rounded-md bg-gray-700 border-gray-600 text-white focus:border-blue-500 focus:ring-blue-500 px-2 py-2"
        id="task-status"
      >
        <option
          value="PENDING"
        >
          Pending
        </option>
        <option
          value="IN_PROGRESS"
        >
          In Progress
        </option>
        <option
          value="COMPLETED"
        >
          Completed
        </option>
      </select>
    </div>
    <div
      class="task-form-field"
    >
      <label
        class="task-form-label block text-sm font-medium text-gray-200 mb-2"
        for="task-due-date"
      >
        Due Date
      </label>
      <input
        class="task-form-input-date mt-1 block w-full rounded-md bg-gray-700 border-gray-600 text-white focus:border-blue-500 focus:ring-blue-500 px-2 py-2"
        id="task-due-date"
        required=""
        type="date"
        value="2024-11-28"
      />
    </div>
    <div
      class="task-form-actions flex justify-end space-x-3 mt-8"
    >
      <button
        class="task-form-button-cancel px-2 py-2 text-sm font-medium text-gray-200 bg-gray-700 border border-gray-600 rounded-md hover:bg-gray-600"
        type="button"
      >
        Cancel
      </button>
      <button
        class="task-form-button-submit px-2 py-2 text-sm font-medium text-white bg-blue-600 border border-transparent rounded-md hover:bg-blue-700"
        type="submit"
      >
        Add Task
      </button>
    </div>
  </form>
</div>
`;
